src/main/scala/cognite/spark/v1/MetricsSource.scala

```diff
@@ -43,9 +62,53 @@ class MetricsSource {
           metrics
         }
       }
-    )
+
+    SparkEnv.get.metricsSystem.registerSource(source)
+
+    source
+  }
+
+  def deregisterMetricSource(key: String, source: Source): Unit = {
+    metricsMap.remove(key)
+    SparkEnv.get.metricsSystem.removeSource(source)
   }
 }
 
 // Singleton to make sure each metric is only registered once.
 object MetricsSource extends MetricsSource
+
+class ExpiringCounter(key: String, source: Source) extends Counter {
+  private var (deathTimer, deathTimerTask): (Timer, TimerTask) = createTimer
+
+  private def createTimer = {
+    val task = new TimerTask {
+      override def run =
+        MetricsSource.deregisterMetricSource(key, source)
+    }
+
+    val timer = new Timer(key)
```

Comment:
here's a thread leak here. Even when we don't have references to TimerThread it will still be considered a garbage collection root by being an alive/running thread

Reply:
Or at least it seems so. There's some code in Timer class that tries to stop the thread when Timer becomes garbage-collectible which might be enough to let the thread stop and be collected, but a bit hard to be really sure this chain would work and no other references path keeps it all alive